core/src/voting_service.rs

```diff
@@ -85,18 +87,49 @@ impl VotingService {
         poh_recorder: Arc<RwLock<PohRecorder>>,
         tower_storage: Arc<dyn TowerStorage>,
         connection_cache: Arc<ConnectionCache>,
+        alpenglow_socket: Option<UdpSocket>,
+        bank_forks: Arc<RwLock<BankForks>>,
     ) -> Self {
         let thread_hdl = Builder::new()
             .name("solVoteService".to_string())
-            .spawn(move || {
-                for vote_op in vote_receiver.iter() {
-                    Self::handle_vote(
-                        &cluster_info,
-                        &poh_recorder,
-                        tower_storage.as_ref(),
-                        vote_op,
-                        connection_cache.clone(),
-                    );
+            .spawn({
+                let mut mock_alpenglow = alpenglow_socket.map(|s| {
+                    MockAlpenglowConsensus::new(
+                        s,
+                        cluster_info.clone(),
+                        EpochSpecs::from(bank_forks.clone()),
+                    )
+                });
+                move || {
+                    for vote_op in vote_receiver.iter() {
+                        // Figure out if we are casting a vote for a new slot, and what slot it is for
+                        let vote_slot = match vote_op {
+                            VoteOp::PushVote {
+                                tx: _,
+                                ref tower_slots,
+                                ..
+                            } => tower_slots.iter().copied().last(),
+                            _ => None,
+                        };
+                        // perform all the normal vote handling routines
+                        Self::handle_vote(
+                            &cluster_info,
+                            &poh_recorder,
+                            tower_storage.as_ref(),
+                            vote_op,
+                            connection_cache.clone(),
+                        );
+                        // trigger mock alpenglow vote if we have just cast an actual vote
+                        if let Some(slot) = vote_slot {
+                            if let Some(ag) = mock_alpenglow.as_mut() {
+                                let root_bank = { bank_forks.read().unwrap().root_bank().clone() };
```

Comment:
nit: I believe root_bank() already does a clone(), so you don't need clone here.

Reply:
Yes it does, thank you!